guides/effectsize_exemplar_within.Rmd

```diff
@@ -31,93 +31,266 @@ format_num <- function(nums, sigdigits = 3) gsub("\\.$", "", formatC(nums, sigdi
+  bootData <- lapply(1:length(bootSubjects), getOneSubjectsData) %>% 
+    bind_rows()
+  
+  # compute the effect sizes the same way we did without bootstrapping
+  afex::aov_ez(
+    data = bootData, 
+    id = 'subject', # subject id column
+    dv = 'response_time', # dependent variable
+    within = c('layout', 'size', 'color'), # within-subject independent variables
+    between = NULL ,# between-subject independent variables
+    #fun_aggregate = mean,
+    anova_table = list(es = 'ges') # effect size = generalized eta squared
+  )$anova_table %>% 
+    as.tibble() %>% 
+    rownames_to_column('effect') %>% # put effect names in a column
+    return()
+}
+```
+
+#### Iterate
+
+The bootstrap needs to run many times to determine how a subset of the data impacts
+
+```{r within-bootstrap-bootstrap iterator}
+
+# run many iterations of the bootstrap procedure
+analyzeManyIterations = function (bootstrapIterationCount) {
+  # each core needs to reload the libraries
+  library(tidyverse)
+  library(afex)
+  lapply(1:bootstrapIterationCount, function(x) analyzeIteration(x)) %>% 
+    bind_rows()
+}
+
+# data used for bootstrapping will collapse by each subject-condition combination
+dataAggregated <- data %>% 
+  group_by(layout, size, color, subject) %>% 
+  summarize(response_time = mean(response_time))
+
 ```
 
-*Note that the fractional degrees of freedom result from a Greenhousse-Geisser sphericity correction.*
 
+#### Parallelize
+
+Bootstrapping can be slow, especially with thousands of iterations. Splitting the iterations across processor cores cuts down on the wait time.
+
+```{r within-parallelize-boostrap, cache=TRUE}
+BOOTSTRAP_COUNT <- 100 # at least 5k recommended. Use lower values for quicker testing
+
+# Initiate cluster
+coreCount <- detectCores() - 1
+ifelse(coreCount < 1, 1, coreCount)
+cl <- makeCluster(coreCount)
+clusterExport(cl, "analyzeIteration")
+clusterExport(cl, "dataAggregated")
+# how many times should each core iterate 
+bootstrapIterationCount <- BOOTSTRAP_COUNT / coreCount
+# run the bootstrap and output the time
+system.time(
+  bootResults <- parLapply(cl, rep(bootstrapIterationCount, coreCount), analyzeManyIterations)
+)
+# the cluster is no longer needed
+stopCluster(cl)
+# cleanup
+rm(coreCount, cl, bootstrapIterationCount, dataAggregated)
 ```
-TODO: Boostrapped 95%CIs for effect sizes
-Pro: people should
-Con: would make the guide even longer
-Maybe push into another guideline?
+
+### Getting a confidence interval from a bootstrap
+
+Each bootstrap iterations ran one anaylsis, so wwe now have many results. So for each effect size, we sort the results and find the range of the inner 95%.
+
+```{r within-boostrap-3}
+# inner 95%
+PERCENTILE_LO <- 0.025
+PERCENTILE_HI <- 0.975
+
+# put all the boostraped results together
+bootResults <- bind_rows(bootResults)
+bootResults <- 
+  bootResults %>% 
+  group_by(effect) %>% 
+  summarize(
+    ESPercentileLo = unname(quantile(ges, probs = PERCENTILE_LO)),
```

Comment:
How about naming these columns `ges_conf_low` and `ges_conf_high`?

Reply:
(see above)